File: fdm-py/1D-heat.cpp
/*C++ version of 1D FDM Heat Equation*/
#include<iostream>
#include<cmath>
#include<vector>
#include<chrono>
#include<algorithm> //for swap(a,b) C++98
#include<fstream>

using namespace std;
using namespace std::chrono;

void theloop(vector<double> fx, vector<double> fxnew,
	     vector<double> peak, vector<double> tminus,
	     int itermax, int imax, double coeff, double dt)
{

}

int main()
{

  vector<double> peak,tminus;
  vector<double> fx, fxnew;

  int itermax = 5000;

  int imax = 100;

  double dx = 1.0/imax;
  double dt = 0.1*dx*dx;
  const double pi = 4.0*atan(1.0);

  double kappa = 1.0;
  double coeff = kappa*dt/(dx*dx);
  
  //peak.resize(1);
  //tminus.resize(1);

  fx.resize(imax);
  fxnew.resize(imax);

  //initiate value for fx
  for(int i=0; i<imax; ++i){
    fx[i] = sin(i*dx*pi);
    fxnew[i] = 0.0;
  }
  
  //initiate peak, tminus
  //peak[0] = 0.0;
  //tminus[0] = 0.0;

  
  high_resolution_clock::time_point t1 = high_resolution_clock::now();
  /*Computation loop*/
  for(int step=0; step<itermax; ++step){
    /*Do something*/
    for(int i=1; i<imax-1; ++i){
      fxnew[i] = fx[i] + coeff*(fx[i-1] -2.0*fx[i]
				+fx[i+1]);
    }
    peak.push_back(fxnew[imax/2]);
    tminus.push_back((step+1)*dt);
    /*Update by swap*/
    swap(fx,fxnew);
  }
  high_resolution_clock::time_point t2 = high_resolution_clock::now();

  duration<double> time_span = duration_cast< duration<double> >(t2-t1);
  double elapsed_time = time_span.count(); //in seconds

  cout<<"Iteration: "<<itermax<<" steps"<<endl;
  cout<<"Time "<<elapsed_time<<" secs"<<endl;

  cout<<fx.size()<<endl;
  
  fstream fp;
  fp.open("decay-cpp.csv", ios::out);
  int row=0;
  fp<<"time, fx_peak\n";
  do{
    fp<<tminus[row]<<", "<<peak[row]<<"\n";
    row +=1;
  }while(row<tminus.size());
  
  fp.close();
  
}

File: fdm-py/1D-heat.py
import numpy as np
import time

pi = np.pi

imax = 100
dx = 1.0/float(imax)

x = np.arange(0.0, 1.0+dx, dx)
#print(x)
#print(x[0])
#print(x[imax])

kappa = 0.25
dt = 0.1*(dx*dx)
coeff = kappa*dt/(dx*dx)

fx = np.sin(pi*x)
fxnew = np.zeros_like(fx)
initfx = fx
#print(fx)

peak = []
tminus = []
icenter = int(imax/2)
peak.append(fx[icenter])
tminus.append(0.0)
#print(peak)
itermax = 5000

ts = time.time()
#iterative loop
for iter in range(itermax):
    for i in range(1,imax):
        fxnew[i] = fx[i] + coeff*( fx[i-1] - 2.0*fx[i] + fx[i+1] )

    #update
    peak.append(fxnew[icenter])
    tminus.append((iter+1)*dt)
    fx = fxnew


ts = time.time() -ts
print("Elapsed time after %d steps: %f"%(itermax, ts))
#print(tminus[:])
    
import matplotlib
import matplotlib.pyplot as plt

fig, ax = plt.subplots()
ax.plot(x, initfx)
ax.plot(x,fxnew)

ax.set(xlabel='x', ylabel='f(x)',
       title='Initial value')
ax.grid()

fig.savefig("test.png")
#plt.show()

#next plot
fig0, decay = plt.subplots()
decay.plot(tminus, peak)

decay.set(xlabel='time', ylabel='f(x)_max',
       title='Peak decay')
decay.grid()

fig0.savefig("decay.png")
